src/sign/participant.rs

```diff
@@ -262,45 +312,326 @@ impl SignParticipant {
+
+        let secret_key = keygen_outputs
+            .iter()
+            .map(|output| bn_to_scalar(output.private_key_share().as_ref()).unwrap())
+            .fold(Scalar::ZERO, |a, b| a + b);
+
+        let r = records[0].x_projection().unwrap();
+
+        let m = <Scalar as Reduce<U256>>::from_be_bytes_reduced(Sha256::digest(message));
+
+        let mut s = k * (m + r * secret_key);
+        s.conditional_assign(&s.negate(), s.is_high());
+
+        let signature = k256::ecdsa::Signature::from_scalars(r, s).unwrap();
+
+        // These checks fail when the overall thing fails
+        let public_key = keygen_outputs[0].public_key().unwrap();
+
+        assert!(public_key.verify(message, &signature).is_ok());
+        assert!(public_key
+            .verify_digest(Sha256::new().chain(message), &signature)
+            .is_ok());
+        signature
+    }
+
+    #[test]
+    fn signing_produces_valid_signature() -> Result<()> {
+        let quorum_size = 4;
+        let rng = &mut init_testing();
+        let sid = Identifier::random(rng);
+
+        // Prepare prereqs for making SignParticipants. Assume all the simulations
+        // are stable (e.g. keep config order)
+        let configs = ParticipantConfig::random_quorum(quorum_size, rng)?;
+        let keygen_outputs = keygen::Output::simulate_set(&configs, rng);
+        let presign_records = PresignRecord::simulate_set(&keygen_outputs, rng);
+
+        let message = b"the quick brown fox jumped over the lazy dog";
+        let message_digest = sha2::Sha256::new().chain(message);
+
+        // Save some things for later -- a signature constructucted from the records and
+        // the public key
+        let non_distributed_sig =
+            compute_non_distributed_ecdsa(message, &presign_records, &keygen_outputs);
+        let public_key = &keygen_outputs[0].public_key().unwrap();
+
+        // Form signing inputs and participants
+        let inputs = std::iter::zip(keygen_outputs, presign_records).map(|(keygen, record)| {
+            sign::Input::new(
+                message_digest.clone(),
+                record,
+                keygen.public_key_shares().to_vec(),
+            )
+        });
+        let mut quorum =
+            std::iter::zip(ParticipantConfig::random_quorum(quorum_size, rng)?, inputs)
+                .map(|(config, input)| {
+                    SignParticipant::new(sid, config.id(), config.other_ids().to_vec(), input)
+                })
+                .collect::<Result<Vec<_>>>()?;
+
+        // Prepare caching of data (outputs and messages) for protocol execution
+        let mut outputs = HashMap::with_capacity(quorum_size);
+
+        let mut inbox = Vec::new();
+        for participant in &quorum {
+            let empty: [u8; 0] = [];
+            inbox.push(Message::new(
+                MessageType::Sign(crate::messages::SignMessageType::Ready),
+                sid,
+                participant.id(),
```

Comment:
I see two parameters to Message::new are both `participant.id()`. Just checking that it is indeed correct?

Reply:
Yeah, ready messages are only sent to yourself. They should not really be `Message`s at all, in my opinion, but we haven't prioritized changing that bit of public API to make it more intuitive. We only recently changed the library to remove the "ready" round, where every party sent a message to every other party to indicate that they were ready to start the protocol (this was never required by the paper) (see #304).

Making a nicer public API for this might be a good next task for you, if you want. The intention of keeping some kind of "ready" message at all was to be able to separate the constructor that makes a new `Participant` from the "start executing protocol" method that gives you the first set of messages to send out to other parties.